### DISCUSSION TITLE
Consoles and output

### DISCUSSION THREAD

**Discussion by kgpierce** [INTO-CPS-Association/into-cps-application#151] [Ideas]
I wanted to open a discussion about how the App displays information from the COE and FMUs during co-simulation. At the moment we have the COE Console which produced minimal information from stdout/stderr (?) and the COE Log which is the information produced during co-simulation.

There are three main issues I see when working with students:

1) Errors can get lost in the COE Log.
This can happen if, for example, there is a run-time error in a VDM-RT FMU which appears in the log, but the co-simulation spits out a few more updates before stopping.  

2) Some warnings get seen as errors when something else is going on.
In particular, "make sure the tool Overture is available" is often quoted as an error. This might change with Maestro 2 as the logging might be different, but still it's something to think about.

3) It can be hard to pick out print statements  
Since there is no way to debug live, students must rely on print statements. These often go flying past during co-simulation and due to the extra data at the start of the line are hard to pick out. It's also not always clear what level of logging override provides (they have moved to INFO but still need to be forced).

I'm open to ideas here, but figured either a way to filter the log or a redesigned log area would be best. My thought around the latter idea would be to have a single Console pane that has tabs like Eclipse, plus the Start / Stop buttons etc. of the current COE Console pane. The main tab would be the Log, potentially with filters on to show / hide ERROR / INFO / DEBUG / TRACE. Additional tabs could be spawned as needed:
- Error (for any warnings or errors, perhaps with filter to hide warnings)
- A named tab for every FMU instance that produces output during co-simulation, e.g. {controller}.controller, {unity3d}.3d, etc.
The individual FMU tabs would show only the output from that FMU, and strip of the "log all OK" etc. to only show the string produced by the model / user.

In this way, users could easily check the output from specific FMUs in real time, to help in things such as watching the console output at the same time as a 3D visualisation to work out what's going on. Additional functionality such as searching, clearing, exporting would be ideal as well.

**Comment by Unsongheroes** [comment]
Hey Ken, this is a really interesting discussion and definitely something that we would like to improve for the users. Based on your proposal I have create a roadmap of sorts, a checklist to be more precise for what and how, we will try and solve this and create a logging system which is more useful, I would like to hear your thoughts on this checklist, and we can modify the list as we go, until we are where we want to be.  
The checklist as it stands is as follows:
- [ ] Create an Initial project, with a few VDM controllers which will be a benchmark project to use for improving the logging. 
- [ ] Get a fuller understanding of the issues at hand.
- [ ] Categorize the issues.
- [ ] Remove unnessassery logs.
- [ ] (Possible) Visually enhance the categorization of the logs.
- [ ] Filtering of the logs.
- [ ] Create tabs for filtering ERROR / INFO / DEBUG / TRACE.
- [ ] (Possible) Creating options for logs (verbose / error ) 
- [ ] Search / clear / export logs.
- [ ] Different FMUs tabs
